tests/eosio.powerup_tests.cpp

```diff
@@ -0,0 +1,858 @@
+#include <Runtime/Runtime.h>
+#include <boost/test/unit_test.hpp>
+#include <cstdlib>
+#include <eosio/chain/contract_table_objects.hpp>
+#include <eosio/chain/exceptions.hpp>
+#include <eosio/chain/global_property_object.hpp>
+#include <eosio/chain/resource_limits.hpp>
+#include <eosio/chain/wast_to_wasm.hpp>
+#include <fc/log/logger.hpp>
+#include <iostream>
+#include <sstream>
+#include <optional>
+#include "eosio.system_tester.hpp"
+
+inline constexpr int64_t powerup_frac  = 1'000'000'000'000'000ll; // 1.0 = 10^15
+inline constexpr int64_t stake_weight = 100'000'000'0000ll; // 10^12
+
+struct powerup_config_resource {
+   std::optional<int64_t>        current_weight_ratio = {};
+   std::optional<int64_t>        target_weight_ratio  = {};
+   std::optional<int64_t>        assumed_stake_weight = {};
+   std::optional<time_point_sec> target_timestamp     = {};
+   std::optional<double>         exponent             = {};
+   std::optional<uint32_t>       decay_secs           = {};
+   std::optional<asset>          min_price            = {};
+   std::optional<asset>          max_price            = {};
+};
+FC_REFLECT(powerup_config_resource,                                                             //
+           (current_weight_ratio)(target_weight_ratio)(assumed_stake_weight)(target_timestamp) //
+           (exponent)(decay_secs)(min_price)(max_price))
+
+struct powerup_config {
+   powerup_config_resource net          = {};
+   powerup_config_resource cpu          = {};
+   std::optional<uint32_t> powerup_days    = {};
+   std::optional<asset>    min_powerup_fee = {};
+};
+FC_REFLECT(powerup_config, (net)(cpu)(powerup_days)(min_powerup_fee))
+
+struct powerup_state_resource {
+   uint8_t        version;
```

Comment:
But we might want to default initialize here.

Reply:
I think we are ok here because that struct is only used from a `get_table`/`fc::raw::unpack`.